libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1065,6 +1065,103 @@ string YulUtilFunctions::clearStorageArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::copyArrayToStorage(ArrayType const& _fromType, ArrayType const& _toType)
+{
+	solAssert(
+		*_fromType.copyForLocation(_toType.location(), _toType.isPointer()).get() == dynamic_cast<ReferenceType const&>(_toType),
+		""
+	);
+
+	string functionName = "copy_array_to_storage_from_" + _fromType.identifier() + "_to_" + _toType.identifier();
+	return m_functionCollector.createFunction(functionName, [&](){
+		Whiskers templ(R"(
+			function <functionName>(slot, <value>) {
+				let length := <arrayLength>(<value>)
+				<?isToDynamic>
+					<storeLength>(slot, length)
+				</isToDynamic>
+
+				for { let i := 0 } lt(i, length) {i := add(i, 1)} {
+					let callDataOrMemoryOffset := <fromArrayIndexAccess>(<value>, i)
```

Comment:
As a reminder. This doesn't work for dynamic nested arrays, as it has length as a return value as well.